spec/mem/tree_borrows/model.md

```diff
@@ -0,0 +1,371 @@
+    allocation: Allocation<TreeBorrowsProvenance>,
+    /// The **TREE** for the Tree Borrows.
+    tree: Tree,
+}
+```
+
+Here we define some helper methods to implement the memory interface.
+
+```rust
+impl<T: Target> TreeBorrowsMemory<T> {
+    /// Given the permission, start address, and the allocation size,
+    /// create an initialized permission map for an allocation.
+    fn init_alloc_permissions(permission: Permission, addr_start: Address, alloc_size: Size) -> Map<Address, Permission> {
+        let mut permissions = Map::new();
+        for addr in addr_start..addr_start + alloc_size.bytes() {
+            permissions.insert(addr, permission);
+        }
+
+        permissions
+    }
+
+    /// Create a new node for a pointer (reborrow)
+    fn new_child(
+        &mut self, 
+        ptr: Pointer<TreeBorrowsProvenance>,
+        pointee_size: Size,
+        permission: Permission
+    ) -> Result<Pointer<TreeBorrowsProvenance>> {
+        // For zero-sized accesses, there is nothing to do.
+        if pointee_size.is_zero() {
+            return ret(ptr);
+        }
+
+        let Some((parent_tag, alloc_id)) = ptr.provenance else {
+            throw_ub!("Tree Borrows: Pointer does not contain provenance");
+        };
+
+        let mut tree_alloc = self.tree_allocs[alloc_id.0];
+        let allocation = tree_alloc.allocation;
+
+        // Create the new child node
+        let child_permissions = Self::init_alloc_permissions(permission, allocation.addr, allocation.size());
+        let child_node = Node {
+            parent: Some(parent_tag),
+            children: List::new(),
+            permissions: child_permissions,
+        };
+
+        let child_tag = self.next_tag();
+        // Add the new node to the parent's children list
+        let Some(mut parent_node) = tree_alloc.tree.nodes.get(parent_tag) else {
+            throw_ub!("Tree Borrows: Parent pointer does not exist in the tree");
+        };
+        parent_node.children.push(child_tag);
+        tree_alloc.tree.insert_node(parent_tag, parent_node);
+        tree_alloc.tree.insert_node(child_tag, child_node);
+
+        // Perform child read to all nodes
+        tree_alloc.tree_access(child_tag, AccessKind::Read, ptr, pointee_size)?;
+
+        self.tree_allocs.set(alloc_id.0, tree_alloc);
+
+        // Create the child pointer and return it 
+        ret(Pointer {
+            provenance: Some((child_tag, alloc_id)),
+            ..ptr
+        })
+    }
+
+    fn next_tag(&mut self) -> BorTag {
+        let tag = self.next_tag;
+        self.next_tag = BorTag(self.next_tag.0 + Int::ONE);
+        tag
+    }
+}
+```
+
+# Memory Operations
+Then we implement the memory model interface for the Tree Borrow.
+
+### Allocate and Deallocate
+
+We create a new tree for one allocation
+
+```rust
+impl<T: Target> Memory for TreeBorrowsMemory<T> {
+    fn allocate(&mut self, kind: AllocationKind, size: Size, align: Align) -> NdResult<Pointer<Self::Provenance>>  {
```

Comment:
This is mostly the same as what the basic memory does, right? Is there a way to share that?

The same goes for `read`, `write`, `deallocate`, and `check_ptr`.